File: raytracing/src/core/light_source.cpp
#include "../../include/core/light_source.h"

LightSource::LightSource(const Vector3 &position, const double intensity) : position(position), intensity(intensity)
{
}

const Vector3 &LightSource::getPosition() const
{
    return position;
}

const LightSource &LightSource::setPosition(const Vector3 &position)
{
    this->position = position;
    return *this;
}

const double LightSource::getIntensity() const
{
    return intensity;
}

const LightSource &LightSource::setIntensity(const double intensity)
{
    this->intensity = intensity;
    return *this;
}

std::ostream &operator<<(std::ostream &os, const LightSource &lightSource)
{
    os << "LightSource(position: " << lightSource.position << ", intensity: " << lightSource.intensity << ")";
    return os;
}

File: src/core/light_source.rs
use crate::{core::Ray, utils::Vector3};

pub trait LightSource: Sync + Send {
    fn get_position(&self) -> &Vector3;
    fn get_intensity(&self) -> f64;
    fn calculate_ray_from_light_source(&self, point: &Vector3) -> Ray;
    fn calculate_lambertian_shading(
        &self,
        point: &Vector3,
        normal: &Vector3,
        albedo: &Vector3,
        light_ray: &Ray,
    ) -> Vector3;
}
